src/ai/monteCarloSearchTree.ts

```diff
@@ -38,4 +57,60 @@ export class MonteCarloSearchTree<T> {
     const randomIndex = Math.floor(Math.random() * node.children.length);
     return node.children[randomIndex];
   }
+
+  rollout(node: Node<T>): void {
+    if (this.isTerminal(node)) {
+      console.log("Result:", this.getValue(node));
+      const reward = this.getValue(node);
+      this.backpropagate(node, reward);
+      return;
+    }
+
+    const options = this.getOptions(node);
+    for (let option of options) {
+      node.addChild(option);
+    }
+    const randomChild = this.selectRandomChild(node);
+    return this.rollout(randomChild);
+  }
+
+  backpropagate(node: Node<T>, reward: number): void {
+    let current;
+    current = node;
+    while (current) {
+      current.visits++;
+      current.score += reward;
+      current = this.getParent(current);
+    }
+  }
+
+  search(iterations: number): Node<T> | undefined {
+    if (iterations < 1) {
+      return this.initialNode.children.sort((a, b) => b.visits - a.visits)[0];
+    }
+    let current = this.current;
+
+    // Get to a leaf node
+    while (!this.isLeafNode(current)) {
+      current = this.getMaxUCB1(current.children);
+    }
+    // Check if leaf node has been visited
+    let hasVisits = this.getVisits(current) > 0;
+
+    if (hasVisits) {
+      current = this.getMaxUCB1(current.children);
+    } else {
+      this.rollout(current);
+    }
+    this.current = current;
+    this.search(iterations - 1);
+  }
 }
+
+const game = new TicTacToeGame(PlayerTypes.human, PlayerTypes.ai);
+const mcts = new MonteCarloSearchTree(
+  new Node(game.getBoardState()),
+  null,
+  game.isGameOver,
+  getReward
+);
```

Comment:
Gotta delete this stuff of course. This whole beast should actually run in a worker since it'll block the main thread.